### PR TITLE
[SPARK-35480][SQL] Make percentile_approx work with pivot

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to avoid wrapping if-else to the constant literals for `percentage` and `accuracy` in `percentile_approx`. They are expected to be literals (or foldable expressions).

Pivot works by two phrase aggregations, and it works with manipulating the input to `null` for non-matched values (pivot column and value).

Note that pivot supports an optimized version without such logic with changing input to `null` for some types (non-nested types basically). So the issue fixed by this PR is only for complex types.

```scala
val df = Seq(
  ("a", -1.0), ("a", 5.5), ("a", 2.5), ("b", 3.0), ("b", 5.2)).toDF("type", "value")
  .groupBy().pivot("type", Seq("a", "b")).agg(
    percentile_approx(col("value"), array(lit(0.5)), lit(10000)))
df.show()
```

**Before:**

```
org.apache.spark.sql.AnalysisException: cannot resolve 'percentile_approx((IF((type <=> CAST('a' AS STRING)), value, CAST(NULL AS DOUBLE))), (IF((type <=> CAST('a' AS STRING)), array(0.5D), NULL)), (IF((type <=> CAST('a' AS STRING)), 10000, CAST(NULL AS INT))))' due to data type mismatch: The accuracy or percentage provided must be a constant literal;
'Aggregate [percentile_approx(if ((type#7 <=> cast(a as string))) value#8 else cast(null as double), if ((type#7 <=> cast(a as string))) array(0.5) else cast(null as array<double>), if ((type#7 <=> cast(a as string))) 10000 else cast(null as int), 0, 0) AS a#16, percentile_approx(if ((type#7 <=> cast(b as string))) value#8 else cast(null as double), if ((type#7 <=> cast(b as string))) array(0.5) else cast(null as array<double>), if ((type#7 <=> cast(b as string))) 10000 else cast(null as int), 0, 0) AS b#18]
+- Project [_1#2 AS type#7, _2#3 AS value#8]
   +- LocalRelation [_1#2, _2#3]
```

**After:**

```
+-----+-----+
|    a|    b|
+-----+-----+
|[2.5]|[3.0]|
+-----+-----+
```

### Why are the changes needed?

To make percentile_approx work with pivot as expected

### Does this PR introduce _any_ user-facing change?

Yes. It threw an exception but now it returns a correct result as shown above.

### How was this patch tested?

Manually tested and unit test was added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -770,6 +770,10 @@ class Analyzer(override val catalogManager: CatalogManager)
                   First(ifExpr(expr), true)
                 case Last(expr, _) =>
                   Last(ifExpr(expr), true)
+                case a: ApproximatePercentile =>
+                  // ApproximatePercentile takes two literals for accuracy and percentage which
+                  // should not be wrapped by if-else.
+                  a.withNewChildren(ifExpr(a.first) :: a.second :: a.third :: Nil)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:776]: So, previously the percentage and accuracy was wrapped by non-foldable if-else which caused analysis failures:

```
if ((type#7 <=> cast(b as string))) 10000 else cast(null as int)
```

File: sql/core/src/test/scala/org/apache/spark/sql/DataFramePivotSuite.scala [hunk 2]
```diff
@@ -344,4 +344,12 @@ class DataFramePivotSuite extends QueryTest with SharedSparkSession {
     val actual = df.groupBy("x").pivot("s").count()
     checkAnswer(actual, expected)
   }
+
+  test("SPARK-35480: percentile_approx should work with pivot") {
+    val actual = Seq(
+      ("a", -1.0), ("a", 5.5), ("a", 2.5), ("b", 3.0), ("b", 5.2)).toDF("type", "value")
+      .groupBy().pivot("type", Seq("a", "b")).agg(
+        percentile_approx(col("value"), array(lit(0.5)), lit(10000)))
+    checkAnswer(actual, Seq((Array(2.5), Array(3.0))).toDF)
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFramePivotSuite.scala:353]: nit: `checkAnswer(actual, Row((Array(2.5), Array(3.0))))` is cheaper?

**Review by maropu — APPROVED** [review]
Nice catch.

**Review by yaooqinn — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1.

**Comment by viirya** [comment]
lgtm

**Comment by dongjoon-hyun** [comment]
Hi, @HyukjinKwon and all.
This broke branch-3.1 compilation. I'll revert this.
```
[error] /home/runner/work/spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:825:46: value first is not a member of org.apache.spark.sql.catalyst.expressions.aggregate.ApproximatePercentile
[error]                   a.withNewChildren(ifExpr(a.first) :: a.second :: a.third :: Nil)
[error]                                              ^
[error] /home/runner/work/spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:825:58: value second is not a member of org.apache.spark.sql.catalyst.expressions.aggregate.ApproximatePercentile
[error]                   a.withNewChildren(ifExpr(a.first) :: a.second :: a.third :: Nil)
[error]                                                          ^
[error] /home/runner/work/spark/spark/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:825:70: value third is not a member of org.apache.spark.sql.catalyst.expressions.aggregate.ApproximatePercentile
[error]                   a.withNewChildren(ifExpr(a.first) :: a.second :: a.third :: Nil)
[error]                                                                      ^
```

**Comment by dongjoon-hyun** [comment]
Since this is not labeled as blocker for Apache Spark 3.1.2, shall we re-try this to land after 3.1.2 RC1?

**Comment by HyukjinKwon** [comment]
Oops sorry @dongjoon-hyun. Yeah this is not a regression or blocker. I'll try after the RC. Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon . And, sorry for the urgent reverting~

**Pull request closed without merging**